### ISSUE TITLE
Feature Request: Access to documents; Score function

### ISSUE THREAD

**Issue by DavZim** [Michael-JB/bm25#2]
Hi Michael,
thank you for the great crate. It works great for me.

There are two small things that would make it even better.
I am mostly using the SearchEngine (following your search example in the readme):

1. access to the underlying hashmap aka document (maybe even the other parts of the struct), alternatively a `get()` trait to return all documents.
2. a score function which returns the score instead of the top K results (eg the `search()` function).

For the second idea, I oftentimes use the following code, which feels a bit clunky:

```rust
use bm25::{Document, Language, SearchEngineBuilder, SearchResult};

let corpus = [
    "The rabbit munched the orange carrot.",
    "The snake hugged the green lizard.",
    "The hedgehog impaled the orange orange.",
    "The squirrel buried the brown nut.",
];

let search_engine = SearchEngineBuilder::<u32>::with_corpus(Language::English, corpus)
    .build();

let mut res: Vec<f32> = vec![0.0; corpus.len()];
let search_results = search_engine.search("orange", corpus.len());

for sr in search_results {
  let id: usize = sr.document.id as usize;
  res[id] = sr.score;
}
// res now holds the scores for each document; 0 if no match is found
```

**Comment by Michael-JB** [comment]
Hi @DavZim, thanks for using the crate, and for your feedback!

I hadn't thought of exposing the scoring in the search engine: the aim was to provide an interface that hides these details. Your suggestion raises a gap in the API of this crate: it exposes embedding and searching, but not scoring. I've not yet come across a use-case where I've needed direct access to a scoring function, so I'm interested in what your use-case is. To help me put together an API that makes the most sense, Is this something you're willing to share?

**Comment by DavZim** [comment]
Hi Michael,
Sure, one of these use cases is hybrid search. Search through a corpus of text using both bm25 as well as vector cosine similarity using embeddings.

**Comment by Michael-JB** [comment]
Hi David,

That's interesting - so you're wanting to implement hybrid search yourself, rather than creating sparse embeddings and deferring the IDF and re-ranking to a vector database, e.g., Qdrant?

For your hybrid search implementation, why do you need scores for all documents (if I've understood your request correctly?). Could you not just do a top-k search in both the sparse and dense spaces and then combine the results with RRF?

**Comment by Michael-JB** [comment]
Hi David,

Thanks for sharing more context - that makes sense. I could imagine an alternative solution in this context is to take the top `k` results from each sub-query, then combine with RRF or similar, and finally take the top `n` results from the combined set (where k >> n).

In either case, I'm currently in the process of some quite significant changes that I hope will make this easier for you. I'll update you here when it's ready :)

**Comment by Michael-JB** [comment]
Hi David,

I've just released v1.0.0, which aligns closer with what I hoped the crate would be when I first started writing it. Included in this release is the new `Scorer` type that lets you (1) score individual document embeddings by their id; and (2) get all matches to a query embedding (i.e., all documents with a non-zero score). You can think of it like an in-memory sparse vector database with a BM25 query API. It also doesn't store the document text: if you're using the scorer directly, you're likely already storing the document text elsewhere so this saves some memory. Also, if you're using the SearchEngine, I've added an `iter` function that returns an iterator over all documents.

I hope that these additions help with your use-case!

See the [release notes](https://github.com/Michael-JB/bm25/releases/tag/v1.0.0) for the full list of changes.
